### PR TITLE
feat: Add type hints on child settings

### PR DESCRIPTION
**Pull request by philasmar**
*Issue #, if available:*
DOTNET-5570
*Description of changes:*
* Add type hints on child settings to enable visual studio toolkit to query the child typehint and return the resources that can be selected. The CLI will still use the parent type hint to display the Object type. The toolkit can decide to use either.
* Added Typehint data for S3BucketName, DynamoDbTable, SNSTopicArns and SQSQueueUrl to choose whether the typehint list allows empty values or not.
* Added a filtering mechanism to retrieve instance types
PR414

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/AWS.Deploy.Orchestration/Data/AWSResourceQueryer.cs [hunk 1]
```diff
@@ -78,6 +81,96 @@ public AWSResourceQueryer(IAWSClientFactory awsClientFactory)
...
+                    }
+                }
+            });
+
+            await foreach (var response in listInstanceTypesPaginator.Responses)
+            {
+                instanceTypes.AddRange(response.InstanceTypes);
+            }
+
+            return instanceTypes.Select(x => (double) x.MemoryInfo.SizeInMiB).Distinct().OrderBy(x => x).ToList();
```
Comment (96malhar) [inline src/AWS.Deploy.Orchestration/Data/AWSResourceQueryer.cs:137]: Bases on the API ref, the `SizeInMiB` is an `int`. Why are we casting it into a double?

API ref - https://docs.aws.amazon.com/sdkfornet/v3/apidocs/items/EC2/TMemoryInfo.html

File: src/AWS.Deploy.Recipes/RecipeDefinitions/ASP.NETAppElasticBeanstalk.recipe [hunk 2]
```diff
@@ -139,11 +140,67 @@
...
+                    ],
+                    "AdvancedSetting": true,
+                    "Updatable": true
+                },
+                {
+                    "Id": "InstanceCPUCores",
+                    "ParentSettingId": "InstanceType",
+                    "Name": "EC2 Instance CPU Cores",
+                    "Description": "The CPU cores of the EC2 instances created for the environment.",
+                    "Type": "Int",
```
Comment (96malhar) [inline src/AWS.Deploy.Recipes/RecipeDefinitions/ASP.NETAppElasticBeanstalk.recipe:175]: Can you add a range validator for the setting and add a validator unit test?

File: src/AWS.Deploy.Recipes/RecipeDefinitions/ASP.NETAppElasticBeanstalk.recipe [hunk 3]
```diff
@@ -139,11 +140,67 @@
...
+                    "TypeHint": "InstanceCPUCores",
+                    "AdvancedSetting": true,
+                    "Updatable": true
+                },
+                {
+                    "Id": "InstanceMemory",
+                    "ParentSettingId": "InstanceType",
+                    "Name": "EC2 Instance Memory",
+                    "Description": "The memory in MB of the EC2 instances created for the environment.",
+                    "Type": "Double",
```
Comment (96malhar) [inline src/AWS.Deploy.Recipes/RecipeDefinitions/ASP.NETAppElasticBeanstalk.recipe:186]: Can you add a range validator for the setting and add a validator unit test?

File: src/AWS.Deploy.Orchestration/Data/AWSResourceQueryer.cs (hunk 1 above)
Comment (philasmar) [reply src/AWS.Deploy.Orchestration/Data/AWSResourceQueryer.cs:137]: In the API ref you linked, it says that it is an Int64 which is mapped to a `long`. I casted the `long` to `double` since we currently support the type `Double` in our recipes.

File: src/AWS.Deploy.Recipes/RecipeDefinitions/ASP.NETAppElasticBeanstalk.recipe (hunk 2 above)
Comment (philasmar) [reply src/AWS.Deploy.Recipes/RecipeDefinitions/ASP.NETAppElasticBeanstalk.recipe:175]: The CPU cores type is dynamically calculated based on the type hint. It is selected from a range of values presented to the end users. It is also not a real option setting item, it is only used to filter the INstance Type. There is no predefined list of values to set as a validator.

File: src/AWS.Deploy.Recipes/RecipeDefinitions/ASP.NETAppElasticBeanstalk.recipe (hunk 3 above)
Comment (philasmar) [reply src/AWS.Deploy.Recipes/RecipeDefinitions/ASP.NETAppElasticBeanstalk.recipe:186]: The instance memory is dynamically calculated based on the type hint. It is selected from a range of values presented to the end users. It is also not a real option setting item, it is only used to filter the Instance Type. There is no predefined list of values to set as a validator.

File: src/AWS.Deploy.Recipes/RecipeDefinitions/ASP.NETAppElasticBeanstalk.recipe (hunk 2 above)
Comment (96malhar) [reply src/AWS.Deploy.Recipes/RecipeDefinitions/ASP.NETAppElasticBeanstalk.recipe:175]: Currently, the type hints are not invoked if a user choses to perform a deployment via a configuration file. It is possible for a user to supply an arbitrarily large value or negative values in the config. Having a simple range validator will throw an actionable error before the CDK hand-off happens.

File: src/AWS.Deploy.Recipes/RecipeDefinitions/ASP.NETAppElasticBeanstalk.recipe (hunk 3 above)
Comment (96malhar) [reply src/AWS.Deploy.Recipes/RecipeDefinitions/ASP.NETAppElasticBeanstalk.recipe:186]: Currently, the type hints are not invoked if a user choses to perform a deployment via a configuration file. It is possible for a user to supply an arbitrarily large value or negative values in the config. Having a simple range validator will throw an actionable error before the CDK hand-off happens.

File: src/AWS.Deploy.Recipes/RecipeDefinitions/ASP.NETAppElasticBeanstalk.recipe (hunk 2 above)
Comment (philasmar) [reply src/AWS.Deploy.Recipes/RecipeDefinitions/ASP.NETAppElasticBeanstalk.recipe:175]: even if the user sets an invalid valid, this does not affect anything. These are filtering OSI, not actual OSI. They don't have a predetermined value, so we cannot create a random range to validate.

File: src/AWS.Deploy.Recipes/RecipeDefinitions/ASP.NETAppElasticBeanstalk.recipe (hunk 3 above)
Comment (philasmar) [reply src/AWS.Deploy.Recipes/RecipeDefinitions/ASP.NETAppElasticBeanstalk.recipe:186]: even if the user sets an invalid valid, this does not affect anything. These are filtering OSI, not actual OSI. They don't have a predetermined value, so we cannot create a random range to validate.

File: src/AWS.Deploy.CLI/Commands/TypeHints/InstanceCPUCoresCommand.cs [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+        {
+            _awsResourceQueryer = awsResourceQueryer;
+            _consoleUtilities = consoleUtilities;
+        }
+
+        private async Task<List<int>?> GetData(Recommendation recommendation, OptionSettingItem optionSetting)
+        {
+            var parentSetting = recommendation.GetOptionSetting(optionSetting.ParentSettingId);
+
+            var freeTierEligible = parentSetting.ChildOptionSettings.FirstOrDefault(x => x.Id.Equals(INSTANCE_FREE_TIER_ELIGIBLE));
```
Comment (normj) [inline src/AWS.Deploy.CLI/Commands/TypeHints/InstanceCPUCoresCommand.cs:31]: We shouldn't have string value type hint that is relying on other sibling settings. What if a recipe wants to use the instance type typehint but it doesn't define an `InstanceFreeTierEligible` or `InstanceArchitecture`

File: src/AWS.Deploy.Recipes/RecipeDefinitions/ASP.NETAppElasticBeanstalk.recipe [hunk 5]
```diff
@@ -139,11 +140,67 @@
...
-            "Type": "String",
+            "Type": "Object",
             "TypeHint": "InstanceType",
-            "DefaultValue": "",
             "AdvancedSetting": true,
-            "Updatable": true
+            "Updatable": true,
+            "ChildOptionSettings": [
+                {
+                    "Id": "InstanceFreeTierEligible",
```
Comment (normj) [inline src/AWS.Deploy.Recipes/RecipeDefinitions/ASP.NETAppElasticBeanstalk.recipe:149]: I don't think this is right to add these settings like InstanceFreeTierEligible, InstanceArchitecture, InstanceCPUCores and InstanceMemory that are not actual settings by filter for the typehint.

File: src/AWS.Deploy.CLI/Commands/TypeHints/S3BucketNameCommand.cs [hunk 6]
```diff
@@ -37,17 +37,15 @@ private async Task<List<S3Bucket>> GetData()
 
         public async Task<object> Execute(Recommendation recommendation, OptionSettingItem optionSetting)
         {
-            const string NO_VALUE = "*** Do not select bucket ***";
             var currentValue = recommendation.GetOptionSettingValue(optionSetting);
             var buckets = (await GetData()).Select(bucket => bucket.BucketName).ToList();
 
-            buckets.Add(NO_VALUE);
```
Comment (normj) [inline src/AWS.Deploy.CLI/Commands/TypeHints/S3BucketNameCommand.cs:44]: Why did you remove the ability to not select a bucket?

File: src/AWS.Deploy.Recipes/RecipeDefinitions/ASP.NETAppElasticBeanstalk.recipe [hunk 7]
```diff
@@ -139,11 +140,67 @@
             "Id": "InstanceType",
             "Name": "EC2 Instance Type",
             "Description": "The EC2 instance type of the EC2 instances created for the environment.",
-            "Type": "String",
+            "Type": "Object",
```
Comment (normj) [inline src/AWS.Deploy.Recipes/RecipeDefinitions/ASP.NETAppElasticBeanstalk.recipe:143]: This is a breaking change to any existing users. I know we are in preview but we should give this serious thought on whether we should be making breaking changes.

**Review by normj — CHANGES_REQUESTED** [review]
The PR should have an animated gif showing the experience of the new type hints. Especially the instance type. As I see the code I'm concerned with the approach.

File: src/AWS.Deploy.CLI/Commands/TypeHints/InstanceCPUCoresCommand.cs (hunk 4 above)
Comment (philasmar) [reply src/AWS.Deploy.CLI/Commands/TypeHints/InstanceCPUCoresCommand.cs]: I removed the extra settings

File: src/AWS.Deploy.Recipes/RecipeDefinitions/ASP.NETAppElasticBeanstalk.recipe (hunk 5 above)
Comment (philasmar) [reply src/AWS.Deploy.Recipes/RecipeDefinitions/ASP.NETAppElasticBeanstalk.recipe]: Done

File: src/AWS.Deploy.CLI/Commands/TypeHints/S3BucketNameCommand.cs (hunk 6 above)
Comment (philasmar) [reply src/AWS.Deploy.CLI/Commands/TypeHints/S3BucketNameCommand.cs:44]: Added typehint data like we discussed

File: src/AWS.Deploy.Recipes/RecipeDefinitions/ASP.NETAppElasticBeanstalk.recipe (hunk 7 above)
Comment (philasmar) [reply src/AWS.Deploy.Recipes/RecipeDefinitions/ASP.NETAppElasticBeanstalk.recipe]: Reverted

**Review by 96malhar — APPROVED** [review]
